piccolo_api/media/azure.py

```diff
@@ -0,0 +1,432 @@
+            Which file extensions are allowed. If ``None``, then all extensions
+            are allowed (not recommended unless the users are trusted).
+        :param allowed_characters:
+            Which characters are allowed in the file name. By default, it's
+            very strict. If set to ``None`` then all characters are allowed.
+        """  # noqa: E501
+        if not azure_import_successful:
+            sys.exit(
+                "Please install azure-storage-blob and azure-identity "
+                "to use this feature `pip install 'piccolo_api[azure]'`"
+            )
+
+        self.storage_account_name = storage_account_name
+        self.container_name = container_name
+        self.upload_metadata = upload_metadata or {}
+        self.folder_name = folder_name
+        self.connection_kwargs = connection_kwargs or {}
+        self.sign_urls = sign_urls
+        self.signed_url_expiry = signed_url_expiry
+        self.executor = executor or ThreadPoolExecutor(max_workers=10)
+
+        self.connection_string = None
+        self.connection_string_parts: t.Dict[str, str] = {}
+        if "connection_string" in self.connection_kwargs:
+            self.connection_string = self.connection_kwargs[
+                "connection_string"
+            ]
+        elif "AZURE_STORAGE_CONNECTION_STRING" in os.environ:
+            self.connection_string = os.environ[
+                "AZURE_STORAGE_CONNECTION_STRING"
+            ]
+        if self.connection_string:
+            self.connection_string_parts = dict(
+                pair.split("=", 1)
+                for pair in self.connection_string.split(";")
+            )
+            if "AccountKey" not in self.connection_string_parts:
+                sys.exit(
+                    "Specified connection string did not contain AccountKey"
+                )
+
+        super().__init__(
+            column=column,
+            allowed_extensions=allowed_extensions,
+            allowed_characters=allowed_characters,
+        )
+
+    def get_client(
+        self,
+    ) -> t.Tuple["ContainerClient", "BlobServiceClient"]:  # pragma: no cover
+        """
+        Returns an Azure client.
+        """
+        if self.connection_string:
+            blob_service_client = BlobServiceClient.from_connection_string(
+                self.connection_string
+            )
+        else:
+            account_url = (
+                f"https://{self.storage_account_name}.blob.core.windows.net"
+            )
+
+            # DefaultAzureCredential is not part of the BlobServiceClient types
+            default_credential = t.cast(str, DefaultAzureCredential())
+
+            blob_service_client = BlobServiceClient(
+                account_url, credential=default_credential
+            )
+
+        container_client = blob_service_client.get_container_client(
+            self.container_name
+        )
+
+        return container_client, blob_service_client
```

Comment:
I don't know a huge amount about Azure, and the differences between the two clients. We could split this into two separate methods ... one for each client type ... not sure. We can leave it as is for now.

Reply:
We could just return the blob_service_client (which we have to instantiate differently based on how you're authing) and then call something like `self.get_client().get_container_client(self.container_name)` if needed but we need the container client 90% of the time, so I figured what we have now is slightly simpler. Up to you.